pkg/blockbuilder/blockbuilder.go

```diff
@@ -437,63 +429,46 @@ consumerLoop:
 	var err error
 	numBlocks, err = builder.CompactAndUpload(ctx, b.uploadBlocks)
 	if err != nil {
-		return err
+		return state, err
 	}
 
-	commitRecOffset := commitRec.Offset + 1 // offset+1 means everything up to (including) the offset was processed
-
-	// We should take the max of "seen till" offset. If the partition was lagging
-	// due to some record not being processed because of a future sample, we might be
-	// coming back to the same consume cycle again.
-	oldestSeenOffset := max(lastRec.Offset, state.OldestSeenOffset)
+	// We should take the max of last seen offsets. If the partition was lagging due to some record not being processed
+	// because of a future sample, we might be coming back to the same consume cycle again.
+	lastSeenOffset := max(lastRec.Offset, state.LastSeenOffset)
 	// Take the max of block max times because of the same reasons above.
 	lastBlockEnd := blockEnd
 	if lastBlockEnd.Before(state.LastBlockEnd) {
 		lastBlockEnd = state.LastBlockEnd
 	}
-
 	commit := kadm.Offset{
 		Topic:       commitRec.Topic,
 		Partition:   commitRec.Partition,
-		At:          commitRecOffset,
+		At:          commitRec.Offset + 1, // offset+1 means everything up to (including) the offset was processed
 		LeaderEpoch: commitRec.LeaderEpoch,
-		Metadata:    marshallCommitMeta(commitRec.Timestamp.UnixMilli(), oldestSeenOffset, lastBlockEnd.UnixMilli()),
+		Metadata:    marshallCommitMeta(commitRec.Timestamp.UnixMilli(), lastSeenOffset, lastBlockEnd.UnixMilli()),
 	}
-	if err := b.commitOffset(ctx, partitionLogger, b.cfg.Kafka.ConsumerGroup, commit); err != nil {
-		return err
+	newState := partitionState{
+		Lag:                state.Lag - (commit.At - firstRec.Offset), // the new lag is the distance between fully processed offsets
```

Comment:
i just realized that offsets aren't guaranteed to be continuous if we use transactions. There may be aborted offsets for which there are no records. Then `Lag` could become negative after doing `state.Lag - (commit.At - firstRec.Offset)`. Isn't this a problem?

I'm not sure what's the best way to deal with this. 1) leave a comment? 2) store the end offset of the partition instead of the Lag and have lag be computed like this.
```go
func (s partitionState) Lag() int64 {
  return s.EndOffset - s.Commit
}
```

In this case Lag shouldn't go below 0.

Reply:
Do we use transactions?

Reply:
we don't

Reply:
Then I think a simple mention in a comment is enough. We don't need to handle an edge for a feature we don't use.

Reply:
> [..] store the end offset of the partition instead of the Lag and have lag be computed like this...

I don't really follow the problem. Lag, as returned by [`kadm.CalculateGroupLagWithStartOffsets`][1] is the difference between the partition's end offset and commit offset, no?

I don't remember the details about the transaction's implementation, but I assumed that the offsets are guaranteed to be monotonic. With that it shouldn't matter if there are gaps between the offset of the previous last commit and the new last commit offset. What do I miss?

[1]: https://pkg.go.dev/github.com/twmb/franz-go/pkg/kadm#CalculateGroupLagWithStartOffsets

Reply:
After brainstorming with @dimitarvdimitrov I've added a comment, that explains the theoretical edge-case and potential way to rework it later https://github.com/grafana/mimir/pull/9199/commits/ac269454239f58e801bf0b06da93d5b8d4bdffa6